charmhelpers/contrib/openstack/context.py

```diff
@@ -2577,6 +2580,23 @@ def cpu_mask(self):
                 mask = mask | 1 << core
         return format(mask, '#04x')
 
+    @classmethod
+    def pmd_cpu_mask(cls):
+        """Get hex formatted pmd CPU mask
+
+        The mask is based on config:pmd-cpu-set.
+        :returns: hex formatted CPU mask
+        :rtype: str
```

Comment:
The function as presented actually has rtype `Optional[str]`, however, do bear in mind that returning that `None` has high risk of blowing up somewhere else, so I'd consider an empty string instead if that makes sense.

Reply:
That would be handled here:
https://github.com/openstack-charmers/charm-layer-ovn/blob/0898177a1f278cb83f911eac70e4a0af7a0dc120/lib/charms/ovn_charm.py#L518-L521

The charm code removes the key from the other_config field when no value is set.

In Python the Boolean check on a variable returns false both for `None` and `''` (the empty string)

Reply:
Just to hijack @fnordahl 's comment a little bit, there *is* an important distinction between `None` (the (no) type) and other types (e.g. `str`).  It specifically is to indicate that *no value* was appropriate, and is intended to either blow up code that doesn't know this (i.e. as a library function return value), or (more correctly) to allow callers to detect the difference between a false-y value and no value.  i.e. we use:

```python
     if some_var is None:
        pass
```

to detect the latter state.

I'm not sure (in the context of this code), which is the better, but @fnordahl is correct in that the *docstring* is incorrect.  The return value *is* Optional[str].

Reply:
used the empty string as suggested by Frode thus kept return type to str.

Reply:
I've changed the behavior in case the option is not set to instead use the dpdk-lcore-mask shifted by 1.